src/rev-address.js

```diff
@@ -0,0 +1,146 @@
+// @ts-check
+import { keccak256 } from 'js-sha3'
+import blake from 'blakejs'
+import { ec } from 'elliptic'
+import { decodeBase16, encodeBase58, encodeBase16, decodeBase58safe } from './lib'
+
+const secp256k1 = new ec('secp256k1')
+
+// Algorithm to generate ETH and REV address is taken from RNode source
+// https://github.com/rchain/rchain/blob/bf7a30e1/rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala#L47
+
+// Prefix as defined in https://github.com/rchain/rchain/blob/c6721a6/rholang/src/main/scala/coop/rchain/rholang/interpreter/util/RevAddress.scala#L13
+const prefix = { coinId : "000000", version: "00" }
+
+/**
+ * @typedef {Object} RevAccount - Represents different formats of REV address
+ * @property {string=} privKey
+ * @property {string=} pubKey
+ * @property {string=} ethAddr
+ * @property {string} revAddr
+ */
+
+/**
+ * Get REV address from ETH address.
+ *
+ * @param {string} ethAddrRaw
+ * @returns {string}
+ */
+export const getAddrFromEth = ethAddrRaw => {
+  const ethAddr = ethAddrRaw.replace(/^0x/, '')
+  if (!ethAddr || ethAddr.length !== 40) return
+
+  // Hash ETH address
+  const ethAddrBytes = decodeBase16(ethAddr)
+  const ethHash      = keccak256(ethAddrBytes)
+
+  // Add prefix with hash and calculate checksum (blake2b-256 hash)
+  const payload      = `${prefix.coinId}${prefix.version}${ethHash}`
+  const payloadBytes = decodeBase16(payload)
+  const checksum     = blake.blake2bHex(payloadBytes, void 666, 32).slice(0, 8)
+
+  // Return REV address
+  return encodeBase58(`${payload}${checksum}`)
+}
+
+/**
+ * Get REV address (with ETH address) from public key.
+ *
+ * @param {string} publicKeyRaw
+ */
+export const getAddrFromPublicKey = publicKeyRaw => {
+  const publicKey = publicKeyRaw.replace(/^0x/, '')
+  if (!publicKey || publicKey.length !== 130) return
+
+  // Public key bytes from hex string
+  const pubKeyBytes = decodeBase16(publicKey)
+  // Remove one byte from pk bytes and hash
+  const pkHash = keccak256(pubKeyBytes.slice(1))
+  // Take last 40 chars from hashed pk (ETH address)
+  const pkHash40 = pkHash.slice(-40)
+
+  // Return both REV and ETH address
+  return {
+    revAddr: getAddrFromEth(pkHash40),
+    ethAddr: pkHash40,
+  }
+}
+
+/**
+ * Get REV address (with ETH address and public key) from private key.
+ *
+ * @param {string} privateKeyRaw
+ */
+export const getAddrFromPrivateKey = privateKeyRaw => {
+  const privateKey = privateKeyRaw.replace(/^0x/, '')
+  if (!privateKey || privateKey.length !== 64) return
+
+  // Generate REV address from private key
+  const key    = secp256k1.keyFromPrivate(privateKey)
+  const pubKey = key.getPublic('hex')
+  const addr   = getAddrFromPublicKey(pubKey)
+
+  // Return public key, REV and ETH address
+  return { pubKey, ...addr }
+}
+
+/**
+ * Verify REV address
+ * @param {string} revAddr
+ */
+export const verifyRevAddr = revAddr => {
+  const revBytes = decodeBase58safe(revAddr)
+  if (!revBytes) return
```

Comment:
again, does this typecheck? it should throw, no? Or change the return type to `string | undefined`.